easyblock = 'PythonBundle'

name = 'Sagemath'
version = '10.3'

homepage = 'https://doc.sagemath.org/html/en/index.html'
description = """Sage is open source mathematical software released under the GNU General Public Licence GPLv2+,
 and includes packages that have compatible software licenses.
 People all around the globe have contributed to the development of Sage. Full documentation is available online."""

toolchain = {'name': 'foss', 'version': '2023b'}

builddependencies = [
    ('Cython', '3.0.9'),
    ('Python-bundle-PyPI', '2023.10'),
    ('pkgconfig', '1.5.5', '-python'),
    ('Boost', '1.83.0'),
]

dependencies = [
    ('Python', '3.11.5'),
    ('SciPy-bundle', '2023.11'),
    ('ECL', '24.5.10'),
    ('FFLAS-FFPACK', '2.5.0'),
    ('eclib', '20231212'),
    ('GSL', '2.7'),
    ('LinBox', '1.7.0'),
    ('Singular', '4.4.0'),
    ('libpng', '1.6.40'),
    ('libgd', '2.3.3'),
    ('m4ri', '20200125'),
    ('cysignals', '1.11.4'),
    ('gmpy2', '2.1.5'),
    ('GMP-ECM', '7.0.5'),
    ('MPFI', '1.5.4'),
    ('cliquer', '1.21'),
    ('nauty', '2.8.8'),
    ('SYMMETRICA', '2.0'),
    ('lcalc', '2.0.5'),
    ('gap', '4.13.0'),
    ('giac', '1.9.0-99'),
    ('m4rie', '20200125'),
    ('libhomfly', '1.02r6'),
    ('libbraiding', '1.2'),
    ('planarity', '3.0.2.0'),
    ('rankwidth', '0.9'),
    ('BRiAl', '1.2.12'),
    ('bliss', '0.77'),
    ('coxeter', '20180226'),
    ('mcqd', '1.0.0'),
    ('SharedMeatAxe', '1.0.1'),
]

use_pip = True
sanity_pip_check = True

exts_list = [
    ('cypari2', '2.1.4', {
        'checksums': ['efa4a44d96f693cb1156da1fd6fccc130dafcf9c19af4185cf770bf44d00dbfc'],
    }),
    ('memory_allocator', '0.1.3', {
        'checksums': ['13805c2ae1c01b7489fab5e8eac9361662b4f2c02412e3652eece48ff6953162'],
    }),
    ('sage-setup', version, {
        'checksums': ['cd5c4e4c9d05abbe8d7cf8dd9b7d78276e06224cd25057745b3c6fc062b1b855'],
    }),
    ('%(namelower)s-standard', version, {
        'checksums': ['43203cd1d1a3c5688c00771b2172af848aca77e3a25055f02928460dec76ef31'],
    }),
    ('%(namelower)s-bliss', version, {
        'checksums': ['41939ecbe2110cd00230152b4761b0ca3a907b48500a52e7e745fe856ee275b3'],
    }),
    ('%(namelower)s-coxeter3', version, {
        'checksums': ['14b51e35a6f8ed0d8f716f81864507fb93b93f2e1fa39e5f86017487d76708ec'],
    }),
    ('%(namelower)s-mcqd', version, {
        'checksums': ['e4b7b91f8e76d58044dedd734abb1cc6333e103d54603e126308f677e354cbb4'],
    }),
    ('%(namelower)s-meataxe', version, {
        'checksums': ['25a4a2da9dafbc6c37975898e5e54e1d477b54bb8bd0554e6a4f88059b382f46'],
    }),
    ('%(namelower)s-sirocco', version, {
        'checksums': ['ff1326fbf30931629d7545c59c4b0c6332802dce6897aced3822658c4a2f9973'],
    }),
    ('%(namelower)s-tdlib', version, {
        'checksums': ['dc7fdec4c29300cf80f5a0ea238434fc1ff89ca2274cc9f96a8c5fd5f0b95eff'],
    }),
]

moduleclass = 'devel'

# TODO
# == 2024-08-27 14:50:00,181 build_log.py:171 ERROR EasyBuild crashed with an error (at easybuild/easybuild-framework/easybuild/base/exceptions.py:126 in __init__): cmd " /apps/gent/RHEL8/cascadelake-ib/software/Python/3.11.5-GCCcore-13.2
# .0/bin/python -m pip install --prefix=/scratch/gent/vo/001/gvo00117/easybuild/RHEL8/cascadelake-ampere-ib/software/Sagemath/10.3-foss-2023b  --no-deps  --ignore-installed  --no-index  --no-build-isolation  ." exited with exit code 1 and
#  output:
# Processing /tmp/vsc45304/easybuild/build/Sagemath/10.3/foss-2023b/sagemathmcqd/sagemath-mcqd-10.3
#   Preparing metadata (pyproject.toml): started
#   Preparing metadata (pyproject.toml): finished with status 'done'
# Building wheels for collected packages: sagemath-mcqd
#   Building wheel for sagemath-mcqd (pyproject.toml): started
#   Building wheel for sagemath-mcqd (pyproject.toml): finished with status 'error'
#   error: subprocess-exited-with-error
#
#    Building wheel for sagemath-mcqd (pyproject.toml) did not run successfully.
#    exit code: 1
#   > [52 lines of output]
#       python_packages = []
#       python_modules = []
#       cython_modules = [<setuptools.extension.Extension('sage.graphs.mcqd') at 0x14873419b390>]
#       [08/27/24 14:49:54] WARNING  pyproject.toml does not contain a tool.setuptools_scm section                                                                                                                                 setuptools.
# py:119
#       running bdist_wheel
#       running build
#       running build_py
#       creating build
#       creating build/lib.linux-x86_64-cpython-311
#       creating build/lib.linux-x86_64-cpython-311/sage
#       creating build/lib.linux-x86_64-cpython-311/sage/graphs
#       copying sage/graphs/all__sagemath_mcqd.py -> build/lib.linux-x86_64-cpython-311/sage/graphs
#       copying sage/graphs/mcqd.pxd -> build/lib.linux-x86_64-cpython-311/sage/graphs
#       running build_ext
#       running build_cython
#       Enabling Cython debugging support
#       INFO: Disabling color, you really want to install colorlog.
#       Disabling color, you really want to install colorlog.
#       Updating Cython code....
#       warning: /scratch/gent/vo/001/gvo00117/easybuild/RHEL8/cascadelake-ampere-ib/software/Sagemath/10.3-foss-2023b/lib/python3.11/site-packages/memory_allocator/memory_allocator.pxd:9:52: Implicit noexcept declaration is deprecated. F
# unction declaration should contain 'noexcept' keyword.
#       warning: sage/graphs/mcqd.pyx:56:19: local variable 'clique_number' referenced before assignment
#       warning: sage/graphs/mcqd.pyx:60:57: local variable 'clique_number' referenced before assignment
#       [1/1] Cythonizing sage/graphs/mcqd.pyx
#       Finished Cythonizing, time: 1.36 seconds.
#       copying ./sage/graphs/mcqd.pxd -> build/lib.linux-x86_64-cpython-311/sage/graphs
#       copying ./sage/graphs/mcqd.pyx -> build/lib.linux-x86_64-cpython-311/sage/graphs
#       building 'sage.graphs.mcqd' extension
#       Executing 1 command (using 1 thread)
#       [1/1] creating build/temp.linux-x86_64-cpython-311/build
#       creating build/temp.linux-x86_64-cpython-311/build/cythonized
#       creating build/temp.linux-x86_64-cpython-311/build/cythonized/sage
#       creating build/temp.linux-x86_64-cpython-311/build/cythonized/sage/graphs
#       gcc -DNDEBUG -g -fwrapv -O3 -Wall -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -O2 -ftree-vectorize -march=native -fno-math-errno -I/apps/gent/RHEL8/cascadelake-ib/software/FFTW/3.3.10-GCC-13.2.0/include -I/apps/gent/RHEL8/cascadelake-ib/software/FlexiBLAS/3.3.1-GCC-13.2.0/include -I/apps/gent/RHEL8/cascadelake-ib/software/FlexiBLAS/3.3.1-GCC-13.2.0/include/flexiblas -fPIC -I/scratch/gent/vo/001/gvo00117/easybuild/RHEL8/cascadelake-ampere-ib/software/cysignals/1.11.4-GCCcore-13.2.0/lib/python3.11/site-packages/cysignals -I/scratch/gent/vo/001/gvo00117/easybuild/RHEL8/cascadelake-ampere-ib/software/Sagemath/10.3-foss-2023b/lib/python3.11/site-packages -I/apps/gent/RHEL8/cascadelake-ib/software/SciPy-bundle/2023.11-gfbf-2023b/lib/python3.11/site-packages/numpy/core/include -I/apps/gent/RHEL8/cascadelake-ib/software/Python/3.11.5-GCCcore-13.2.0/include/python3.11 -Ibuild/cythonized -I/apps/gent/RHEL8/cascadelake-ib/software/Python/3.11.5-GCCcore-13.2.0/include/python3.11 -c build/cythonized/sage/graphs/mcqd.cpp -o build/temp.linux-x86_64-cpython-311/build/cythonized/sage/graphs/mcqd.o -fno-strict-aliasing -DCYTHON_CLINE_IN_TRACEBACK=1
#       build/cythonized/sage/graphs/mcqd.cpp:1223:10: fatal error: mcqd.h: No such file or directory
#        1223 | #include "mcqd.h"
#             |          ^~~~~~~~
#       compilation terminated.
#       error: command '/apps/gent/RHEL8/cascadelake-ib/software/GCCcore/13.2.0/bin/gcc' failed with exit code 1
#       Exception ignored in: <function Pool.__del__ at 0x1486ea34f740>
#       Traceback (most recent call last):
#         File "/apps/gent/RHEL8/cascadelake-ib/software/Python/3.11.5-GCCcore-13.2.0/lib/python3.11/multiprocessing/pool.py", line 271, in __del__
#           self._change_notifier.put(None)
#         File "/apps/gent/RHEL8/cascadelake-ib/software/Python/3.11.5-GCCcore-13.2.0/lib/python3.11/multiprocessing/queues.py", line 377, in put
#           self._writer.send_bytes(obj)
#         File "/apps/gent/RHEL8/cascadelake-ib/software/Python/3.11.5-GCCcore-13.2.0/lib/python3.11/multiprocessing/connection.py", line 199, in send_bytes
#           self._send_bytes(m[offset:offset + size])
#         File "/apps/gent/RHEL8/cascadelake-ib/software/Python/3.11.5-GCCcore-13.2.0/lib/python3.11/multiprocessing/connection.py", line 410, in _send_bytes
#           self._send(header + buf)
#         File "/apps/gent/RHEL8/cascadelake-ib/software/Python/3.11.5-GCCcore-13.2.0/lib/python3.11/multiprocessing/connection.py", line 367, in _send
#           n = write(self._handle, buf)
#               ^^^^^^^^^^^^^^^^^^^^^^^^
#       OSError: [Errno 9] Bad file descriptor
#       [end of output]
#
#   note: This error originates from a subprocess, and is likely not a problem with pip.
#   ERROR: Failed building wheel for sagemath-mcqd
# Failed to build sagemath-mcqd
# ERROR: Could not build wheels for sagemath-mcqd, which is required to install pyproject.toml-based projects
#  (at easybuild/easybuild-framework/easybuild/tools/run.py:682 in parse_cmd_output)
